OptimizelySDK/Odp/LruCache.cs

```diff
@@ -0,0 +1,171 @@
+﻿/* 
+ * Copyright 2022, Optimizely
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+using OptimizelySDK.Logger;
+using OptimizelySDK.Utils;
+using System;
+using System.Collections.Generic;
+
+namespace OptimizelySDK.Odp
+{
+    public class LruCache<T> : ICache<T>
+        where T : class
+    {
+        private readonly int _maxSize;
+        private readonly object _mutex;
+        private readonly TimeSpan _timeout;
+        private readonly TimeSpan _timeoutDisabled = TimeSpan.Zero;
+        private readonly ILogger _logger;
+        private readonly Dictionary<string, (LinkedListNode<string> node, ItemWrapper value)> _cache;
+        private readonly LinkedList<string> _list;
+
+        public LruCache(int? maxSize = null,
+            TimeSpan? itemTimeout = null, ILogger logger = null
+        )
+        {
+            const int DEFAULT_MAX_SIZE = 10000;
+            const int CACHE_DISABLED = 0;
+            var defaultTimeout = TimeSpan.FromMinutes(10);
+
+            _mutex = new object();
+
+            if (maxSize is null)
+            {
+                _maxSize = DEFAULT_MAX_SIZE;
+            }
+            else if (maxSize < 0)
+            {
+                _maxSize = CACHE_DISABLED;
+            }
+            else
+            {
+                _maxSize = maxSize.Value;
+            }
+
+            if (itemTimeout is null)
+            {
+                _timeout = defaultTimeout;
+            }
+            else if (itemTimeout?.TotalMilliseconds < 0)
+            {
+                _timeout = _timeoutDisabled;
+            }
+            else
+            {
+                _timeout = itemTimeout.Value;
+            }
+
+            _logger = logger ?? new DefaultLogger();
+
+            _cache =
+                new Dictionary<string, (LinkedListNode<string> node, ItemWrapper value)>(_maxSize);
```

Comment:
Am just curious can't we just take LinkedListNode<ItemWrapper> as value?

Reply:
I think the idea is to pop off both the node from the Tuple and put it in the `_list` & back in the `_cache` along with the `item`  see linw 150ish.

Let's talk through this too.